### ISSUE TITLE
Add ability to "save/load" docker images without pushing

### ISSUE THREAD

**Issue by midN** [bazelbuild/rules_k8s#177]
Hello,

It would be nice to have a flag to just save/load image from tarball locally without actually pushing it.

Scenario:
We want to run k8s resources locally for development purposes.
For that case we need to build/save image locally from tarball and not pushing it to repo.

Currently we have to run separate bazel targets in a shell wrapper, one to build images/save and one to run k8s resources.
Ideally i should be able to provide single `k8s_object(s)` with `images` attribute which would build the images, save them locally and run k8s resources

**Comment by steeve** [comment]
A little ping for this one, as it makes an amazing partner to a local k8s environment for development. Personally I use it also with ksonnet and rules_jsonnet and the workflow is great.

**Comment by steeve** [comment]
I'm guessing this is because of a limitation of `rules_docker` which allows to create and publish images _without having docker_. Doing the required `docker load` would require docker to be installed.

**Comment by chrislovecnm** [comment]
Is this a duplicate #223 ish?

**Comment by steeve** [comment]
Loosely yes, but with this issue, the image still needs to be imported via `docker load`, since the use case is using local k8s cluster (minikube or docker for mac).

**Comment by chrislovecnm** [comment]
So here Is my understanding. You can create and push containers w/o docker. You can run a container w/o docker.

Do it all the time 😊

Where does the toolchain die?

**Comment by steeve** [comment]
In my case I'm using docker for mac + the kubernetes integration. In this case, there is no need for a registry since all the images are already downloaded.

What I'm trying to do is use rules_k8s+rules_docker+rules_jsonnet+ksonnet for my local development environment.

**Comment by chrislovecnm** [comment]
Right so you want the manifest generated, but the container built but not staged.

**Comment by steeve** [comment]
Exactly. If it's already build on the local docker instance, it will be available to the local k8s cluster (in the case of docker for mac).

So not only needs to be built, but also imported and available via `docker images`.

The simplest thing being using `docker load`, I think.

**Comment by chrislovecnm** [comment]
To be clear you want the k8s manifest built but 

https://github.com/bazelbuild/rules_k8s/blob/master/k8s/resolver.py

The resolver does not push the image and kubectl is not applied.

Basically a run target for just generating the manifest. Is that what you are trying to do? I am also surprised that bazel build does not render the manifest.

**Comment by steeve** [comment]
I do want `kubectl` to be called.
However, it should basically run `docker load image.tar.gz` instead of pushing the image to the registry.
And indeed use call `kubectl` to run the container.

One thing that might not be obvious is that in the case of Docker for Windows/Mac, a local k8s context is created so that all kubectl commands run, only on the local cluster.

**Comment by chrislovecnm** [comment]
So that is different 😁

So you want a “docker import” and you want k8s templates rendered. 

Also what k8s environment are you using? Last time I used minikube I needed a registry, but that was long ago.

Also not sure why we are using that python rather than bazel containers load.

**Comment by steeve** [comment]
Indeed a custom resolver might do the trick !

**Comment by chrislovecnm** [comment]
Why don’t you just use a local registry running on k8s?

**Comment by steeve** [comment]
I'm using Docker for Mac's own kubernetes integration: https://blog.docker.com/2018/01/docker-mac-kubernetes/

Also, running a local registry should pretty wasteful with the image living twice on the disk, I think?

**Comment by chrislovecnm** [comment]
Not saying no, just not certain of the proper or best design.

Also it would live twice on disk anyways. Once for the build and once in docker storage path. But we are talking mb. Unless you have 1gb containers.

We need other people’s feed back on the design.

Running a local registry would work for you now.

**Comment by steeve** [comment]
Indeed you're right about the living twice.
Thinking about it, why is doing a custom `docker load` resolver (could live outside this repo) a bad design?

**Comment by nlopezgi** [comment]
@alex1545 fyi

**Comment by abatilo** [comment]
Can we get some traction back here? I would love to be able to run a bazel build stack the same way I might use [skaffold](https://github.com/GoogleContainerTools/skaffold) or [tilt](https://github.com/tilt-dev/tilt) to doing local development with a local Kubernetes as my running environment.

**Comment by hipstern** [comment]
Basically we want to ability to take bazel image targets and load them into the local docker daemon as to avoid having to push images onto a remote registry. this is common practice with tools such as skaffold, etc.

**Comment by aakarim** [comment]
To chime in here, if you `bazel run` a `container_image` target right now it automatically pushes it to your local docker registry.
